File: backend/app/services/extraction.py
"""LLM-based extraction service using MiniMax (OpenAI-compatible API)."""
import json
import logging
import re
from typing import Optional

from openai import (
    AsyncOpenAI,
    OpenAI,
    AuthenticationError,
    APIConnectionError,
    RateLimitError,
)

from app.core.config import settings

logger = logging.getLogger(__name__)

_SYSTEM_PROMPT = (
    "You are an academic paper analyst. Given a paper's metadata, "
    "extract the requested information.\n\n"
    "Rules:\n"
    "1. Respond ONLY with a single JSON object, nothing else.\n"
    "2. Do NOT include any explanation, reasoning, or thinking.\n"
    '3. Format: {"value": "concise answer", "evidence": "quote from abstract"}\n'
    '4. If not available: {"value": null, "evidence": null}\n'
    "5. Use your knowledge about well-known papers to supplement the abstract."
)


def _build_user_prompt(paper_data: dict, column_prompt: str) -> str:
    """Build the user prompt from paper metadata and column prompt.

    Args:
        paper_data: Dict with title, abstract, authors, year.
        column_prompt: The extraction instruction for this column.

    Returns:
        Formatted user prompt string.
    """
    title = paper_data.get("title", "Unknown")
    year = paper_data.get("year", "Unknown")
    authors = paper_data.get("authors", [])
    abstract = paper_data.get("abstract", "No abstract available.")

    if isinstance(authors, list):
        authors_str = ", ".join(authors)
    else:
        authors_str = str(authors)

    return (
        f"Paper: {title} ({year})\n"
        f"Authors: {authors_str}\n"
        f"Abstract: {abstract}\n\n"
        f"Extract: {column_prompt}"
    )


def _get_client() -> AsyncOpenAI:
    """Create an AsyncOpenAI client configured for MiniMax.

    Returns:
        Configured AsyncOpenAI client.
    """
    return AsyncOpenAI(
        api_key=settings.MINIMAX_API_KEY,
        base_url=settings.MINIMAX_BASE_URL,
    )


class ExtractionError(Exception):
    """Raised when LLM API call fails (auth, network, rate-limit, etc.)."""

    def __init__(self, message: str) -> None:
        self.message = message
        super().__init__(message)


async def extract_cell_with_llm(
    paper_data: dict,
    column_prompt: str,
) -> Optional[dict]:
    """Extract a single cell value from paper metadata using LLM.

    Args:
        paper_data: Dict with title, abstract, authors, year.
        column_prompt: The extraction instruction for this column.

    Returns:
        Dict with 'value' and 'evidence_text' keys, or None if the
        information is not available in the paper.

    Raises:
        ExtractionError: When the LLM API call fails.
    """
    client = _get_client()
    user_prompt = _build_user_prompt(paper_data, column_prompt)

    try:
        response = await client.chat.completions.create(
            model=settings.MINIMAX_MODEL,
            messages=[
                {"role": "system", "content": _SYSTEM_PROMPT},
                {"role": "user", "content": user_prompt},
            ],
            temperature=0.1,
            max_tokens=512,
        )
    except AuthenticationError as e:
        logger.error("LLM authentication failed: %s", e)
        raise ExtractionError("LLM API key invalid or expired") from e
    except RateLimitError as e:
        logger.error("LLM rate limit hit: %s", e)
        raise ExtractionError("LLM API rate limit exceeded, please retry later") from e
    except APIConnectionError as e:
        logger.error("LLM connection failed: %s", e)
        raise ExtractionError("Cannot connect to LLM API") from e
    except Exception as e:
        logger.exception("LLM extraction failed for paper: %s", paper_data.get("title"))
        raise ExtractionError(f"LLM API error: {e}") from e

    try:
        content = response.choices[0].message.content
        if not content:
            logger.warning("Empty response from LLM for paper: %s", paper_data.get("title"))
            return None

        # Strip <think>...</think> blocks (reasoning models)
        content = content.strip()
        content = re.sub(r"<think>[\s\S]*?</think>", "", content).strip()

        # Strip markdown code fences if present
        if content.startswith("```"):
            content = content.split("\n", 1)[-1]
            if content.endswith("```"):
                content = content[:-3]
            content = content.strip()

        # If content is empty after stripping (model only produced thinking),
        # return None
        if not content:
            logger.warning(
                "LLM response was empty after stripping think tags for paper: %s",
                paper_data.get("title"),
            )
            return None

        result = json.loads(content)

        if result.get("value") is None:
            return None

        return {
            "value": str(result["value"]),
            "evidence_text": result.get("evidence") or "",
        }

    except json.JSONDecodeError:
        logger.error(
            "Failed to parse LLM response as JSON for paper: %s, raw: %s",
            paper_data.get("title"),
            content,
        )
        return None


# ========== Sync extraction for thread pool ==========

_sync_client: Optional[OpenAI] = None


def _get_sync_client() -> OpenAI:
    """Get or create a thread-safe sync OpenAI client.

    Returns:
        Configured sync OpenAI client.
    """
    global _sync_client
    if _sync_client is None:
        _sync_client = OpenAI(
            api_key=settings.MINIMAX_API_KEY,
            base_url=settings.MINIMAX_BASE_URL,
        )
    return _sync_client


def _parse_llm_content(content: Optional[str], title: str) -> Optional[dict]:
    """Parse LLM response content into structured result.

    Args:
        content: Raw LLM response content.
        title: Paper title for logging.

    Returns:
        Dict with 'value' and 'evidence_text', or None.
    """
    if not content:
        logger.warning("Empty response from LLM for paper: %s", title)
        return None

    content = content.strip()
    content = re.sub(r"<think>[\s\S]*?</think>", "", content).strip()

    if content.startswith("```"):
        content = content.split("\n", 1)[-1]
        if content.endswith("```"):
            content = content[:-3]
        content = content.strip()

    if not content:
        logger.warning("LLM response empty after stripping for paper: %s", title)
        return None

    try:
        result = json.loads(content)
    except json.JSONDecodeError:
        logger.error("Failed to parse LLM JSON for paper: %s, raw: %s", title, content)
        return None

    if result.get("value") is None:
        return None

    return {
        "value": str(result["value"]),
        "evidence_text": result.get("evidence") or "",
    }


def extract_cell_sync(paper_data: dict, column_prompt: str) -> Optional[dict]:
    """Sync extraction for use in ThreadPoolExecutor.

    Args:
        paper_data: Dict with title, abstract, authors, year.
        column_prompt: The extraction instruction for this column.

    Returns:
        Dict with 'value' and 'evidence_text', or None.

    Raises:
        ExtractionError: When the LLM API call fails.
    """
    client = _get_sync_client()
    user_prompt = _build_user_prompt(paper_data, column_prompt)

    try:
        response = client.chat.completions.create(
            model=settings.MINIMAX_MODEL,
            messages=[
                {"role": "system", "content": _SYSTEM_PROMPT},
                {"role": "user", "content": user_prompt},
            ],
            temperature=0.1,
            max_tokens=512,
        )
    except AuthenticationError as e:
        logger.error("LLM authentication failed: %s", e)
        raise ExtractionError("LLM API key invalid or expired") from e
    except RateLimitError as e:
        logger.error("LLM rate limit hit: %s", e)
        raise ExtractionError("LLM API rate limit exceeded, please retry later") from e
    except APIConnectionError as e:
        logger.error("LLM connection failed: %s", e)
        raise ExtractionError("Cannot connect to LLM API") from e
    except Exception as e:
        logger.exception("LLM extraction failed for paper: %s", paper_data.get("title"))
        raise ExtractionError(f"LLM API error: {e}") from e

    return _parse_llm_content(
        response.choices[0].message.content,
        paper_data.get("title", "Unknown"),
    )

File: lib/server/extraction.ts
/**
 * LLM-based extraction service using MiniMax / OpenAI-compatible API
 * (replaces backend/app/services/extraction.py)
 */
import OpenAI from 'openai'
import { config } from './config'

const SYSTEM_PROMPT = [
  'You are an academic paper analyst. Given a paper\'s metadata, ',
  'extract the requested information.\n\n',
  'Rules:\n',
  '1. Respond ONLY with a single JSON object, nothing else.\n',
  '2. Do NOT include any explanation, reasoning, or thinking.\n',
  '3. Format: {"value": "concise answer", "evidence": "quote from abstract"}\n',
  '4. If not available: {"value": null, "evidence": null}\n',
  '5. Use your knowledge about well-known papers to supplement the abstract.',
].join('')

function buildUserPrompt(
  paperData: Record<string, unknown>,
  columnPrompt: string,
): string {
  const title = (paperData.title as string) || 'Unknown'
  const year = (paperData.year as string | number) || 'Unknown'
  const rawAuthors = paperData.authors
  let authorsStr: string
  if (Array.isArray(rawAuthors)) {
    authorsStr = rawAuthors.join(', ')
  } else {
    authorsStr = String(rawAuthors || '')
  }
  const abstract =
    (paperData.abstract as string) || 'No abstract available.'

  return (
    `Paper: ${title} (${year})\n` +
    `Authors: ${authorsStr}\n` +
    `Abstract: ${abstract}\n\n` +
    `Extract: ${columnPrompt}`
  )
}

let client: OpenAI | null = null

function getClient(): OpenAI {
  if (!client) {
    client = new OpenAI({
      apiKey: config.MINIMAX_API_KEY,
      baseURL: config.MINIMAX_BASE_URL,
    })
  }
  return client
}

export class ExtractionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ExtractionError'
  }
}

function repairTruncatedJson(
  raw: string,
): { value: string; evidence: string } | null {
  const valueMatch = raw.match(/"value"\s*:\s*"((?:[^"\\]|\\.)*)"/)
  if (valueMatch) {
    const value = valueMatch[1]
    const evidenceMatch = raw.match(
      /"evidence"\s*:\s*"((?:[^"\\]|\\.)*)"/,
    )
    const evidence = evidenceMatch ? evidenceMatch[1] : ''
    return { value, evidence }
  }
  return null
}

function parseLlmContent(
  content: string | null | undefined,
): { value: string; evidence_text: string } | null {
  if (!content) return null

  let cleaned = content.trim()

  // Strip <think>...</think> blocks
  cleaned = cleaned.replace(/<think>[\s\S]*?<\/think>/g, '').trim()

  // Strip markdown code fences
  if (cleaned.startsWith('```')) {
    cleaned = cleaned.split('\n').slice(1).join('\n')
    if (cleaned.endsWith('```')) {
      cleaned = cleaned.slice(0, -3)
    }
    cleaned = cleaned.trim()
  }

  if (!cleaned) return null

  let result: Record<string, unknown>
  try {
    result = JSON.parse(cleaned)
  } catch {
    const repaired = repairTruncatedJson(cleaned)
    if (!repaired) return null
    result = repaired
  }

  if (result.value == null) return null

  return {
    value: String(result.value),
    evidence_text: (result.evidence as string) || '',
  }
}

export async function extractCellWithLlm(
  paperData: Record<string, unknown>,
  columnPrompt: string,
): Promise<{ value: string; evidence_text: string } | null> {
  const llm = getClient()
  const userPrompt = buildUserPrompt(paperData, columnPrompt)

  let response: OpenAI.Chat.Completions.ChatCompletion
  try {
    response = await llm.chat.completions.create({
      model: config.MINIMAX_MODEL,
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: userPrompt },
      ],
      temperature: 0.1,
      max_tokens: 1024,
    })
  } catch (err: unknown) {
    if (err instanceof OpenAI.AuthenticationError) {
      throw new ExtractionError('LLM API key invalid or expired')
    }
    if (err instanceof OpenAI.RateLimitError) {
      throw new ExtractionError(
        'LLM API rate limit exceeded, please retry later',
      )
    }
    if (err instanceof OpenAI.APIConnectionError) {
      throw new ExtractionError('Cannot connect to LLM API')
    }
    throw new ExtractionError(`LLM API error: ${err}`)
  }

  return parseLlmContent(response.choices[0]?.message?.content)
}
